service/src/java/org/apache/hive/service/servlet/OTELExporter.java

```diff
@@ -0,0 +1,229 @@
+
+            while (task.getEndTime() == null) {
+              try {
+                Thread.sleep(500);
+              } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+              }
+            }
+            currSpan.end(task.getEndTime(), TimeUnit.MILLISECONDS);
+          }
+        }
+
+        if(rootspan != null){
+          queryIdToSpanMap.put(queryID,rootspan);
+          queryIdToTasksMap.put(queryID,completedTasks);
+        }
+
+      }
+    }
+
+
+    for (QueryInfo hQuery : historicalQueries) {
+
+      //For queries that were live till last loop but have ended before start of this loop
+      if(queryIdToSpanMap.containsKey(hQuery.getQueryDisplay().getQueryId())){
+        String hQueryId = hQuery.getQueryDisplay().getQueryId();
+        Span rootspan = queryIdToSpanMap.get(hQueryId);
+        for (QueryDisplay.TaskDisplay task : hQuery.getQueryDisplay().getTaskDisplays()) {
+          if(queryIdToTasksMap.get(hQueryId) == null || !queryIdToTasksMap.get(hQueryId).contains(task.getTaskId())){
+            queryIdToTasksMap.get(hQueryId).add(task.getTaskId());
+            Context parentContext = Context.current().with(rootspan);
+            Span currSpan = tracer.spanBuilder(hQueryId+ " - " + task.getTaskId() + " - completed").setParent(parentContext).setAllAttributes(addTaskAttributes(task))
+                    .setStartTimestamp(task.getBeginTime(), TimeUnit.MILLISECONDS).startSpan();
+            currSpan.end(task.getEndTime(), TimeUnit.MILLISECONDS);
+          }
+        }
+
+        //Update the rootSpan name & attributes before ending it
+        queryIdToSpanMap.get(hQueryId).updateName(hQueryId + " - completed");
+        queryIdToSpanMap.get(hQueryId).setAllAttributes(addQueryAttributes(hQuery));
+        queryIdToSpanMap.get(hQueryId).end();
+
+        queryIdToSpanMap.remove(hQueryId);
+        queryIdToTasksMap.remove(hQueryId);
+
+        historicalQueryId.add(hQueryId);
+      }
+
+      //For queries that were already either before OTEL service started or in between OTEL loops
+      if (!historicalQueryId.contains(hQuery.getQueryDisplay().getQueryId())) {
+        historicalQueryId.add(hQuery.getQueryDisplay().getQueryId());
+        Span rootSpan = tracer.spanBuilder(hQuery.getQueryDisplay().getQueryId() + " - completed")
+                .startSpan();
+        for (QueryDisplay.TaskDisplay task : hQuery.getQueryDisplay().getTaskDisplays()) {
+          Context parentContext = Context.current().with(rootSpan);
+          Span currSpan = tracer.spanBuilder(hQuery.getQueryDisplay().getQueryId()+ " - " + task.getTaskId() + " - completed").setParent(parentContext).setAllAttributes(addTaskAttributes(task))
+                  .setStartTimestamp(task.getBeginTime(), TimeUnit.MILLISECONDS).startSpan();
+          currSpan.end(task.getEndTime(), TimeUnit.MILLISECONDS);
+        }
+        rootSpan.setAllAttributes(addQueryAttributes(hQuery)).end();
+      }
+
```

Comment:
Your `historicalQueryId` will grow infinitely, you need to reset it the only ones returned by ``    List<QueryInfo> historicalQueries = operationManager.getHistoricalQueryInfos();``

This would return say only 25 (by default) but you would be storing it infinitely which will ultimately lead to OOM on the HS2 side

Reply:
Yes, i have refactored the code such that at the end of each OTEL loop, all the queries which have been removed from `historicalQueries` will also be removed from our `historicalQueryId`.